### PR TITLE
Add job_name as a staticmethod in SLURMEnvironment class

### LINKED ISSUE #10697: Add job_name as a staticmethod in SLURMEnvironment class
## Proposed refactor

Add job_name as a staticmethod in SLURMEnvironment class to improve code quality of the staticmethod job_id:
https://github.com/PyTorchLightning/pytorch-lightning/blob/6fc7c54c3af6dc3a446253df379e459d18bacb85/pytorch_lightning/plugins/environments/slurm_environment.py#L41-L54

### Motivation
Currently:
```python
os.environ.get("SLURM_JOB_NAME")
```
is used to get the slurm job name. This could be replaced by a staticmethod in the SLURMEnvironment class to improve the code quality.

### Pitch

1. Add a staticmethod job_name to the SLURMEnvironment class.
2. Replace os.environ.get("SLURM_JOB_NAME") with the staticmethod job_name in the staticmethod job_id.

### Additional context

@awaelchli 

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta

### PR DESCRIPTION
**Pull request by AndresAlgaba**
## What does this PR do?
Fixes #10697, and moved the staticmethod job_id (out of the properties) for readability of the code.

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by AndresAlgaba** [comment]
Should any new tests be added? And should the changelog be updated for this refactoring?

**Comment by awaelchli** [comment]
Looks great! Adding a single line here (assert None): 

https://github.com/PyTorchLightning/pytorch-lightning/blob/dca1776870e9dd7236c03d7ccd55fcb09a9bb5e7/tests/plugins/environments/test_slurm_environment.py#L24

and one here with a fictional value for a job name

https://github.com/PyTorchLightning/pytorch-lightning/blob/dca1776870e9dd7236c03d7ccd55fcb09a9bb5e7/tests/plugins/environments/test_slurm_environment.py#L52

should do it.

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/environments/slurm_environment.py [hunk 1]
```diff
@@ -100,6 +85,25 @@ def detect() -> bool:
...
+            try:
+                job_id = int(job_id)
+            except ValueError:
+                job_id = None
+
+        # in interactive mode, don't make logs use the same job id
+        in_slurm_interactive_mode = SLURMEnvironment.job_name() == "bash"
+        if in_slurm_interactive_mode:
+            job_id = None
+        return job_id
```
Comment (ananthsub) [inline pytorch_lightning/plugins/environments/slurm_environment.py:105]: @awaelchli why do we do this logic inside of the job id call instead of at the log time?

Comment (awaelchli) [reply pytorch_lightning/plugins/environments/slurm_environment.py:105]: yes that's what I meant in the comment https://github.com/PyTorchLightning/pytorch-lightning/pull/10622#discussion_r752622231 over here, as we can disentangle and simplify this. 
The code was already like this before we moved it to SLURMEnvironment.

Comment (AndresAlgaba) [reply pytorch_lightning/plugins/environments/slurm_environment.py:105]: Should we do it in this PR?

Comment (awaelchli) [reply pytorch_lightning/plugins/environments/slurm_environment.py:105]: No, because this PR is focusing on adding the job_name, but the comment is appreciated and definitely an opportunity for a new PR.

File: pytorch_lightning/plugins/environments/slurm_environment.py [hunk 2]
```diff
@@ -100,6 +85,25 @@ def detect() -> bool:
...
+        return os.environ.get("SLURM_JOB_NAME")
+
+    @staticmethod
+    def job_id() -> Optional[int]:
+        job_id = os.environ.get("SLURM_JOB_ID")
+        if job_id:
+            try:
+                job_id = int(job_id)
+            except ValueError:
+                job_id = None
```
Comment (Borda) [inline pytorch_lightning/plugins/environments/slurm_environment.py:99]:
```suggestion
        try:
            job_id = int(job_id)
        except ValueError:
            job_id = None
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/plugins/environments/slurm_environment.py [hunk 3]
```diff
@@ -92,11 +92,10 @@ def job_name() -> Optional[str]:
...
         job_id = os.environ.get("SLURM_JOB_ID")
-        if job_id:
-            try:
-                job_id = int(job_id)
-            except ValueError:
-                job_id = None
+        try:
+            job_id = int(job_id)
+        except ValueError:
+            job_id = None
```
Comment (ananthsub) [inline pytorch_lightning/plugins/environments/slurm_environment.py]: 1. you could return early if `job_id is None` to avoid this try/catch...
2. you could move the check from L101 to happen first. if the job_name is `"bash"` then also return `None` early. 
```suggestion
        if job_id is None:
            return None
        try:
            job_id = int(job_id)
        except ValueError:
            job_id = None
```

Comment (AndresAlgaba) [reply pytorch_lightning/plugins/environments/slurm_environment.py:108]: Thanks for the feedback! I've implemented these changes in the new "change order" commit.
